File: examples/quantified/z3py/memmem.py
from z3 import *
s = Solver()

i = Int('i')
j = Int('j')
needle_len = Int('needle_len')
haystack_len = Int('haystack_len')
ret = Int('ret')

I = IntSort()
needle = Array('needle', I, I)
haystack = Array('haystack', I, I)

s.add(
	Implies(ret == haystack_len, 
		ForAll(i, 
			Implies(i < haystack_len - needle_len, 
				Exists(j, 
					And(And(0 <= j, j < needle_len),
						Select(needle, j) != Select(haystack, i + j)
					)
				)
			)
		)
	)
)

s.add(
	Implies(ret < haystack_len,
		And(
			ForAll(i, 
				Implies(
					And(0 <= i, i < needle_len), 
					Select(haystack, ret + i) == Select(needle, i)
				)
			), 
			ForAll(i, 
				Implies(i < ret, 
					Exists(j, 
						And(
							And(0 <= j, j < needle_len),
							Select(haystack, j + i) != Select(needle, j)
						)
					)
				)
			)
		)
	)
)

s.add(haystack_len == 5)
s.add(needle_len == 2)
s.add(ret == 2)

# should fail on this
# s.add(Select(haystack, 0) == Select(needle, 0)) 
# s.add(Select(haystack, 1) == Select(needle, 1)) 

# should fail on this
# s.add(Select(haystack, 1) == Select(needle, 0)) 
# s.add(Select(haystack, 2) == Select(needle, 1)) 


s.add(Select(haystack, 2) == Select(needle, 0))
s.add(Select(haystack, 3) == Select(needle, 1))


print(s.check())
print(s.model())

File: examples/quantified/memmem/memmem.c
#include "klee/klee.h"
#include <stdio.h>

char* match(char* haystack, char* needle, uint32_t haystack_len, uint32_t needle_len) {
        uint32_t ret;
        klee_make_symbolic(&ret, sizeof(ret), "memmem_return_value");
        klee_memmem(haystack, needle, haystack_len, needle_len, ret);
        return haystack + ret;
}

int main() {
        char haystack[5], needle[2];
        uint32_t haystack_len = 5, needle_len = 2;
        klee_make_symbolic(haystack, haystack_len, "haystack");
        klee_make_symbolic(needle, needle_len, "needle");
        char* res = match(haystack, needle, haystack_len, needle_len);
        
        
        // Test #0: match found on index 2 - check for no occurences before that
        // klee_assume(res == haystack + 2);
        // klee_assume(haystack[1] == needle[0]); 
        // klee_assume(haystack[2] == needle[1]); // should fail on this
        

        
        // Test #1: match found - check if match is actually found
        // klee_assume(res == haystack + 2);
        // klee_assume(haystack[2] == needle[0]);
        // klee_assume(haystack[3] != needle[1]); // should fail on this
        

       
        // Test #2: match not found 
        // klee_assume(res == haystack + haystack_len);
        // klee_assume(haystack[2] == needle[0]);
        // klee_assume(haystack[3] == needle[1]);
}
